### ISSUE TITLE
[Native] reduce_agg returns incorrect results 

### ISSUE THREAD

**Issue by mknegi** [prestodb/presto#21177]
This query returns incorrect results if ran against c++ worker

```
SELECT
    x,
    ARRAY_JOIN(
        ARRAY_SORT(
            REDUCE_AGG(y, ARRAY['x'], (a, b) -> a || b, (a, b) -> a || b)
        ),
        ''
    )
FROM (
    VALUES
        (1, ARRAY['a']),
        (1, ARRAY['b']),
        (1, ARRAY['c']),
        (2, ARRAY['d']),
        (2, ARRAY['e']),
        (3, ARRAY['f'])
) AS t(x, y)
GROUP BY
    x
```

Presto:

```
 x | _col1
---+-------
 2 | dex
 1 | abcx
 3 | fx
(3 rows)
```

Velox:

```
Actual rows (up to 100 of 2 extra rows shown, 3 rows in total):
  [2, dexx]
  [1, abcxxx]
Expected rows (up to 100 of 2 missing rows shown, 3 rows in total):
  [1, abcx]
  [2, dex]
```

References:
- https://prestodb.io/docs/current/functions/aggregate.html#reduce_agg
- https://velox-lib.io/blog/reduce-agg

**Comment by aditi-pandit** [comment]
This does seem like a difference in behavior between reduce_agg in Velox vs Presto.

Presto applies the input function only once for each group, whereas Velox applies the input function for each input row in the group (so we get as many x's as there are rows in the group).

@mbasmanova : Wanted to bring this on your radar. What do you think ?

**Comment by mbasmanova** [comment]
I believe this is an incorrect usage of reduce_agg function. This function is rather tricky to use.

The problem is that aggregations need to support running distributed. Let's consider 2 cases. Both should return same results. In the first scenario, all rows are processed on a single node. In the second, each row is first partially aggregated on a separate machines, then all combined for final aggregation to another machine.

Input:

```
---+-----
 1 | [a]
 1 | [b]
 2 | [c]
 2 | [d]
```

In the first case, the results in Presto will be bax or abx (depending on the order of inputs) + dcx or cdx.

In the second case, the results of partial agg will be (1 row per node):

```
---+-----
 1 | [ax]
 1 | [bx]
 2 | [cx]
 2 | [dx]
```

And final agg will produce: axbx or bxax + cxdx or dxcx. 

You can see that results would be highly affected by how distributed the query run. Not what users expect.

A proper way to write this reduce_agg would be:

```
    ARRAY['x'] || ARRAY_JOIN(
        ARRAY_SORT(
            REDUCE_AGG(y, ARRAY[], (a, b) -> a || b, (a, b) -> a || b)
        ),
        ''
    )
```

In general, the initialState and combineFunction needs to satisfy this criteria: combineFunction(initialState, initialState) = initialState. Or, more generally,

- combine(input(x, initialState), input(y, initialState)) = input(y, input(x, initialState))
- combine(s1, s2) = combine(s2, s1)

for any x, y, s1 and s2.

**Comment by mbasmanova** [comment]
I suggest to update documentation for reduce_agg to clarify the requirements for the inputs.

**Comment by mbasmanova** [comment]
Here is a PR to improve reduce_agg documentation in Velox. A similar change needs to go in for Presto docs.

https://github.com/facebookincubator/velox/pull/7160

**Comment by yzhang1991** [comment]
@steveburnett 👆

**Comment by yzhang1991** [comment]
@mknegi Can you help update the native test for this particular test case to accept the new result?
@pramodsatya Please help clarify with Manoj.

**Comment by aditi-pandit** [comment]
@mbasmanova : Thanks for the detailed explanation. This is really helpful.

Had a follow up question : It sounds like we could do some optimizer changes as well (it sounds like previously the partials would need to be on a single node) in line for this.

Is that understanding fair ?

**Comment by mbasmanova** [comment]
@aditi-pandit Aditi, I'm not sure I understand the question. Would you clarify a bit?

**Comment by aditi-pandit** [comment]
@mbasmanova : If I do EXPLAIN (type distributed) on the query I get a SINGLE fragment. Should this change to a multi-fragment, distributed plan because of how Velox has implemented reduce_agg now ?

```
presto:tiny> explain (type distributed) SELECT x, ARRAY_JOIN( ARRAY_SORT( REDUCE_AGG(y, ARRAY['x'], (a, b) -> a || b, (a, b) -> a || b) ), '' ) FROM ( VALUES (1, ARRAY['a']), (1, ARRAY['b']), (1, ARRAY['c']), (2, ARRAY['d']), (2, ARRAY['e']), (3, ARRAY['f']) ) AS t(x, y) GROUP BY x;
                                                                                      Query Plan                                            >
-------------------------------------------------------------------------------------------------------------------------------------------->
 Fragment 0 [SINGLE]                                                                                                                        >
     Output layout: [field, array_join]                                                                                                     >
     Output partitioning: SINGLE []                                                                                                         >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                          >
     - Output[x, _col1] => [field:integer, array_join:varchar]                                                                              >
             x := field (1:35)                                                                                                              >
             _col1 := array_join (1:38)                                                                                                     >
         - Project[projectLocality = LOCAL] => [field:integer, array_join:varchar]                                                          >
                 array_join := array_join(array_sort(reduce_agg), VARCHAR'') (1:38)                                                         >
             - Aggregate(FINAL)[field][$hashvalue] => [field:integer, $hashvalue:bigint, reduce_agg:array(varchar(1))]                      >
                     reduce_agg := "presto.default.reduce_agg"((reduce_agg_18,(expr, expr_0) -> concat(expr, expr_0),(expr_1, expr_2) -> con>
                 - LocalExchange[HASH][$hashvalue] (field) => [field:integer, reduce_agg_18:array(varchar(1)), $hashvalue:bigint]           >
                     - Aggregate(PARTIAL)[field][$hashvalue_19] => [field:integer, $hashvalue_19:bigint, reduce_agg_18:array(varchar(1))]   >
                             reduce_agg_18 := "presto.default.reduce_agg"((field_3,expr_10,(expr, expr_0) -> concat(expr, expr_0),(expr_1, e>
                         - Project[projectLocality = LOCAL] => [field:integer, field_3:array(varchar(1)), expr_10:array(varchar(1)), $hashva>
                                 Estimates: {rows: 6 (360B), cpu: 1476.00, memory: 0.00, network: 0.00}                                     >
                                 expr_10 := [Block: position count: 1; size: 168 bytes]                                                     >
                             - LocalExchange[ROUND_ROBIN] () => [field:integer, field_3:array(varchar(1)), $hashvalue_19:bigint]            >
                                     Estimates: {rows: 6 (360B), cpu: 780.00, memory: 0.00, network: 0.00}                                  >
                                 - Project[projectLocality = LOCAL] => [field:integer, field_3:array(varchar(1)), $hashvalue_20:bigint]     >
                                         Estimates: {rows: 6 (360B), cpu: 390.00, memory: 0.00, network: 0.00}                              >
                                         $hashvalue_20 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(field), BIGINT'0')) (1:139)  >
                                     - Values => [field:integer, field_3:array(varchar(1))]                                                 >
                                             Estimates: {rows: 6 (360B), cpu: 0.00, memory: 0.00, network: 0.00}                            >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'2', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'2', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'3', [Block: position count: 1; size: 168 bytes])                                      >
```

**Comment by mbasmanova** [comment]
> Can you help update the native test for this particular test case to accept the new result?

Just to clarify. The update should be to change initialState argument to ARRAY[] (empty array).

**Comment by mbasmanova** [comment]
Also, you may need to add sorting to make results deterministic:

```
array_sort(REDUCE_AGG(y, ARRAY[], (a, b) -> a || b, (a, b) -> a || b))
```

or 

```
REDUCE_AGG(y, ARRAY[], (a, b) -> array_sort(a || b), (a, b) -> array_sort(a || b))
```

**Comment by aditi-pandit** [comment]
> @aditi-pandit Aditi, I'm not sure I understand the question. Would you clarify a bit?

@mbasmanova : If I do EXPLAIN (type distributed) on the query I get a SINGLE fragment. Should this change to a multi-fragment, distributed plan because of how Velox has implemented reduce_agg now ?

```
presto:tiny> explain (type distributed) SELECT x, ARRAY_JOIN( ARRAY_SORT( REDUCE_AGG(y, ARRAY['x'], (a, b) -> a || b, (a, b) -> a || b) ), '' ) FROM ( VALUES (1, ARRAY['a']), (1, ARRAY['b']), (1, ARRAY['c']), (2, ARRAY['d']), (2, ARRAY['e']), (3, ARRAY['f']) ) AS t(x, y) GROUP BY x;
                                                                                      Query Plan                                            >
-------------------------------------------------------------------------------------------------------------------------------------------->
 Fragment 0 [SINGLE]                                                                                                                        >
     Output layout: [field, array_join]                                                                                                     >
     Output partitioning: SINGLE []                                                                                                         >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                          >
     - Output[x, _col1] => [field:integer, array_join:varchar]                                                                              >
             x := field (1:35)                                                                                                              >
             _col1 := array_join (1:38)                                                                                                     >
         - Project[projectLocality = LOCAL] => [field:integer, array_join:varchar]                                                          >
                 array_join := array_join(array_sort(reduce_agg), VARCHAR'') (1:38)                                                         >
             - Aggregate(FINAL)[field][$hashvalue] => [field:integer, $hashvalue:bigint, reduce_agg:array(varchar(1))]                      >
                     reduce_agg := "presto.default.reduce_agg"((reduce_agg_18,(expr, expr_0) -> concat(expr, expr_0),(expr_1, expr_2) -> con>
                 - LocalExchange[HASH][$hashvalue] (field) => [field:integer, reduce_agg_18:array(varchar(1)), $hashvalue:bigint]           >
                     - Aggregate(PARTIAL)[field][$hashvalue_19] => [field:integer, $hashvalue_19:bigint, reduce_agg_18:array(varchar(1))]   >
                             reduce_agg_18 := "presto.default.reduce_agg"((field_3,expr_10,(expr, expr_0) -> concat(expr, expr_0),(expr_1, e>
                         - Project[projectLocality = LOCAL] => [field:integer, field_3:array(varchar(1)), expr_10:array(varchar(1)), $hashva>
                                 Estimates: {rows: 6 (360B), cpu: 1476.00, memory: 0.00, network: 0.00}                                     >
                                 expr_10 := [Block: position count: 1; size: 168 bytes]                                                     >
                             - LocalExchange[ROUND_ROBIN] () => [field:integer, field_3:array(varchar(1)), $hashvalue_19:bigint]            >
                                     Estimates: {rows: 6 (360B), cpu: 780.00, memory: 0.00, network: 0.00}                                  >
                                 - Project[projectLocality = LOCAL] => [field:integer, field_3:array(varchar(1)), $hashvalue_20:bigint]     >
                                         Estimates: {rows: 6 (360B), cpu: 390.00, memory: 0.00, network: 0.00}                              >
                                         $hashvalue_20 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(field), BIGINT'0')) (1:139)  >
                                     - Values => [field:integer, field_3:array(varchar(1))]                                                 >
                                             Estimates: {rows: 6 (360B), cpu: 0.00, memory: 0.00, network: 0.00}                            >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'1', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'2', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'2', [Block: position count: 1; size: 168 bytes])                                      >
                                             (INTEGER'3', [Block: position count: 1; size: 168 bytes])                                      >
```

**Comment by yzhang1991** [comment]
With the updates to the documentation, we will close this as a WontFix
